src/app/reporting/Engine.cpp

```diff
@@ -94,59 +93,69 @@ Engine::RetrieveClusterData(FabricIndex aAccessingFabricIndex, AttributeReportIB
         {
-            // Retrieve data for this cluster instance and clear its dirty flag.
-            err = RetrieveClusterData(apReadHandler->GetFabricIndex(), AttributeReportIBs, *clusterInfo);
-            VerifyOrExit(err == CHIP_NO_ERROR,
-                         ChipLogError(DataManagement, "<RE:Run> Error retrieving data from cluster, aborting"));
-            attributeClean = false;
-        }
-        else
-        {
-            for (auto path = mpGlobalDirtySet; path != nullptr; path = path->mpNext)
+            if (!apReadHandler->IsInitialReport())
             {
-                if (clusterInfo->IsAttributePathSupersetOf(*path))
-                {
-                    // SetDirty injects path into GlobalDirtySet path that don't have the particular nodeId,
-                    // need to inject nodeId from subscribed path here.
-                    ClusterInfo dirtyPath = *path;
-                    dirtyPath.mNodeId     = clusterInfo->mNodeId;
-                    err                   = RetrieveClusterData(apReadHandler->GetFabricIndex(), AttributeReportIBs, dirtyPath);
-                }
-                else if (path->IsAttributePathSupersetOf(*clusterInfo))
+                bool concretePathDirty = false;
+                // TODO: Optimize this implementation by making the iterator only emit intersected paths.
+                for (auto dirtyPath = mpGlobalDirtySet; dirtyPath != nullptr; dirtyPath = dirtyPath->mpNext)
                 {
-                    err = RetrieveClusterData(apReadHandler->GetFabricIndex(), AttributeReportIBs, *clusterInfo);
+                    if (dirtyPath->IsPathIncluded(path))
+                    {
+                        concretePathDirty = true;
+                        break;
+                    }
                 }
-                else
+
+                if (!concretePathDirty)
                 {
-                    // partial overlap is not possible, hence the 'continue' here: clusterInfo and path have nothing in
-                    // common.
+                    // This attribute is not dirty, we just skip this one.
                     continue;
                 }
-                VerifyOrExit(err == CHIP_NO_ERROR,
-                             ChipLogError(DataManagement, "<RE:Run> Error retrieving data from cluster, aborting"));
-                attributeClean = false;
             }
+
+            TLV::TLVWriter attributeBackup;
+            attributeReportIBs.Checkpoint(attributeBackup);
+            // Retrieve data for this cluster instance and clear its dirty flag.
+            err = RetrieveClusterData(apReadHandler->GetFabricIndex(), attributeReportIBs, path);
+            if (err != CHIP_NO_ERROR)
+            {
+                attributeReportIBs.Rollback(attributeBackup);
+            }
+            SuccessOrExit(err);
+            attributeDataWritten = true;
         }
+        mMoreChunkedMessages = false;
```

Comment:
This variable isn't actually used to track state across multiple message transmissions, so making it a member variable gives a false impression of that purpose. It would be better to just pass that as in-out parameters in these function calls to better indicate it's more narrowly scoped usage.